SolarTracker/Pages/Index.razor

```diff
@@ -0,0 +1,362 @@
+        </GraphicsLayer>
+    </Map>
+    <Graphic>
+        <Point Longitude="_center?.Longitude" Latitude="_center?.Latitude" />
+    </Graphic>
+    <LocateWidget @ref="_locateWidget" Position="OverlayPosition.TopLeft" />
+    <SearchWidget Position="OverlayPosition.TopRight" />
+</MapView>
+
+@code {
+
+    protected override async Task OnInitializedAsync()
+    {
+        await FetchTimeZoneData();
+    }
+
+    private async Task OnMapRendered()
+    {
+        if (!_terminusAdded)
+        {
+            await GetCenterAndTimeZone();
+            await CreateTerminator();
+            _terminusAdded = true;
+            await UpdateGraphic();
+        }
+        else if (_cancel)
+        {
+            _cancel = false;
+            _movingToSunrise = false;
+            _movingToSunset = false;
+        }
+        else if (_movingToSunrise)
+        {
+            if (!await GeometryEngine.Intersects(_center!, _polygon!))
+            {
+                await MoveTimeUntilStateChanged((int)(_skipInterval / 2.0), false, false);
+                _movingToSunrise = false;
+            }
+            else
+            {
+                _selectedDateTime = _selectedDateTime.AddMinutes(_skipInterval);
+                await CreateTerminator();
+            }
+            await UpdateGraphic();
+        }
+        else if (_movingToSunset)
+        {
+            if (await GeometryEngine.Intersects(_center!, _polygon!))
+            {
+                await MoveTimeUntilStateChanged((int)(_skipInterval / 2.0), false, true);
+                _movingToSunset = false;
+            }
+            else
+            {
+                _selectedDateTime = _selectedDateTime.AddMinutes(_skipInterval);
+                await CreateTerminator();
+            }
+            await UpdateGraphic();
+        }
+    }
+
+    private async Task CreateTerminator()
+    {
+        foreach (var graphic in _graphicsLayer!.Graphics)
+        {
+            await _graphicsLayer.Remove(graphic);
+        }
+        
+        int ordinalDay = _selectedDateTime.DayOfYear;
+        double solarDeclination = -57.295779 * Math.Asin(0.397788 * Math.Cos(0.017203 * (ordinalDay + 10) + 0.052465 * Math.Sin(0.017203 * (ordinalDay - 2))));
+        SpatialReference spatialReference = (await _mapView!.GetSpatialReference())!;
+        
+        bool isWebMercator = spatialReference.Wkid is null ||
+                             (int)spatialReference.Wkid == 102100 ||
+                             (int)spatialReference.Wkid == 3857 ||
+                             (int)spatialReference.Wkid == 102113;
+        double yMax = isWebMercator ? 85 : 90;
+        double latitude = yMax * (solarDeclination > 0 ? -1 : 1);
+        List<MapPath> rings = new();
+        DateTime utcDateTime = _selectedDateTime.Subtract(_timeZoneOffset);
+        
+        for (double lon = -180; lon < 180; lon++)
+        {
+            MapPath path = new(new(lon + 1, latitude), 
```

Comment:
I'm not sure that I follow what's happening here. You are building a stepped terminator line, at one-degree increments, but each step is a separate ring? I don't think I would normally use Rings in that way, normally rings represent separate "parts" of a geometry, not parts of a continuous line.

If I put a`break;` right after `rings.Add(path)` such that it only draws one ring, I get a big line, and that seems not right, but I also could be super wrong here.
<img width="832" alt="image" src="https://user-images.githubusercontent.com/1805118/205835707-8f2ea4e9-3542-4fa7-9926-7c0d6acd0c97.png">

Reply:
The calculations are all from https://github.com/jgravois/midnight-commander, which I will add to the comments. I would not have known how to generate this at all otherwise. If you have a better suggestion, I'm open to it.

Reply:
I finally dug into this a bit, and the calculation seem correct, but it is drawing a bunch of small rectangular shapes to create the terminator polygon. (It's using a polygon so that one side of it, the polygon side, is filled in). I did a little experimenting and you can do it without creating 365 rings in the polygon. This doesn't really matter, and I can't tell that it is actually any faster, but here is the code for posterity:

```csharp
        var line = new MapPath();
        for (double lon = -180; lon < 180; lon++)
        {
            line.Add(new(lon, GetLatitude(lon, solarDeclination, -yMax, yMax, utcDateTime)));
            line.Add(new(lon + 1, GetLatitude(lon, solarDeclination, -yMax, yMax, utcDateTime)));
        }
        // add the fixed line to make the full polygon
        line.Add(new (180, GetLatitude(180, solarDeclination, -yMax, yMax, utcDateTime)));
        line.Add(new (180, latitude));
        line.Add(new (-180, latitude));
        line.Add(new (-180, GetLatitude(-180, solarDeclination, -yMax, yMax, utcDateTime)));
```